server/etcdserver/healthz.go

```diff
@@ -0,0 +1,397 @@
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/server/v3/auth"
+)
+
+const (
+	PathLivez   = "/livez"
+	PathReadyz  = "/readyz"
+	PathHealthz = "/healthz"
+)
+
+type EtcdServerHealth interface {
+	Alarms() []*etcdserverpb.AlarmMember
+	Range(context.Context, *etcdserverpb.RangeRequest) (*etcdserverpb.RangeResponse, error)
+}
+
+// HealthChecker is a named healthz checker.
+type HealthChecker interface {
+	Name() string
+	Check() error
+}
+
+type healthLogger interface {
+	Infof(template string, args ...interface{})
+}
+
+func (s *EtcdServer) InstallLivezReadyz(lg *zap.Logger, mux mux) {
+	s.healthHandler.installLivez(lg, mux)
+	s.healthHandler.installReadyz(lg, mux)
+	s.healthHandler.installHealthz(lg, mux)
+}
+
+func (s *EtcdServer) AddHealthCheck(check HealthChecker, isLivez bool, isReadyz bool) error {
+	return s.healthHandler.addHealthCheck(check, isLivez, isReadyz)
+}
+
+type HealthHandler struct {
+	server EtcdServerHealth
```

Comment:
I feel like HealthHandler should be in the api layer and move to `server/etcdserver/api/`. If it is only used for etcdhttp and move it there.

Reply:
Even though the api change is not desirable in 3.5, are you sure we don't need to add it to gPRC in the future? Existing health check has different/repeated logic at different places, which is quite confusing.

Reply:
moved to server/etcdserver/api/v3health